From dbe0e1f03213ebf2b52bbb44336ed30ca158e51b Mon Sep 17 00:00:00 2001
From: David Daney <ddaney@caviumnetworks.com>
Date: Tue, 27 Jan 2009 14:31:01 -0800
Subject: [PATCH] boot cpu interrupt affinity fix

1) Set the default interrupt affinity to the boot CPU.

2) Set correct processor affinity for watchdog interrupts.

Signed-off-by: David Daney <ddaney@caviumnetworks.com>
Signed-off-by: Phil Staub <Phil.Staub@windriver.com>
---
 arch/mips/cavium-octeon/watchdog.c |    8 +++++++-
 arch/mips/kernel/irq-octeon.c      |    4 ++++
 2 files changed, 11 insertions(+), 1 deletions(-)

diff --git a/arch/mips/cavium-octeon/watchdog.c b/arch/mips/cavium-octeon/watchdog.c
index 5331f70..9f77878 100644
--- a/arch/mips/cavium-octeon/watchdog.c
+++ b/arch/mips/cavium-octeon/watchdog.c
@@ -197,11 +197,17 @@ static void setup_interrupt(void *arg)
 	cvmx_ciu_wdogx_t ciu_wdog;
 	int core = cvmx_get_core_num();
 	unsigned long threshold = (unsigned long) arg;
+	unsigned int irq = OCTEON_IRQ_WDOG0 + core;
 
-	if (request_irq(OCTEON_IRQ_WDOG0 + core, watchdog_poke_irq, IRQF_SHARED,
+	if (request_irq(irq, watchdog_poke_irq, IRQF_SHARED,
 		    "watchdog", watchdog_poke_irq)) {
+		panic("watchdog couldn't obtain irq %u", irq);
 	}
 
+#ifdef CONFIG_SMP
+	/* Set the affinity to this cpu. */
+	irq_set_affinity(irq, cpumask_of_cpu(smp_processor_id()));
+#endif
 	/* Poke the watchdog to clear out its state */
 	cvmx_write_csr(CVMX_CIU_PP_POKEX(core), 1);
 
diff --git a/arch/mips/kernel/irq-octeon.c b/arch/mips/kernel/irq-octeon.c
index 738f3c6..fd47c7d 100644
--- a/arch/mips/kernel/irq-octeon.c
+++ b/arch/mips/kernel/irq-octeon.c
@@ -415,6 +415,10 @@ void __init arch_init_irq(void)
 {
 	int irq;
 
+#ifdef CONFIG_SMP
+	/* Set the default affinity to the boot cpu. */
+	irq_default_affinity = cpumask_of_cpu(smp_processor_id());
+#endif
 	if (NR_IRQS < OCTEON_IRQ_LAST)
 		pr_err("octeon_irq_init: NR_IRQS is set too low\n");
 
-- 
1.6.0.4

